### ISSUE TITLE
Talk 7: Spenser Bauman

### ISSUE THREAD

**Issue by rrnewton** [wonks/ICFP_rehearsal_feedback#7]
Practice talk -- 23:10 total time.

**Comment by ezyang** [comment]
"we could go the other direction": took me a moment to figure out what you meant here (this is the contract slide)

**Comment by cgswords** [comment]
- What does `for/sum` do?
- Remove 'background' slide. Put something better up there.

**Comment by cfbolz** [comment]
Liked the motivation and introductory examples

**Comment by rrnewton** [comment]
CEK machine slide rides the bottom of the slide closely -- it looks like a bit was cut off on our projector and you may have the same problem on other projectors.

**Comment by cgswords** [comment]
All of these slides, I think, with single words in the middle, should go away.

The CEK machine code is... lengthy. I don't think the full reduction semantics adds a ton.

**Comment by rrnewton** [comment]
my 2 cents: I personally don't mind the one word slides... reminds of Matt Might's (and Lawrence Lessig's) minimal style.

**Comment by jasonhemann** [comment]
Also, Steve Jobs.

**Comment by cgswords** [comment]
Yeah, Jobs was my thought.

E: Spenser should just get a turtleneck?

**Comment by jasonhemann** [comment]
Some of the larger words seemed to cast a shadow behind them. If it's an artifact of the projector, then c'est la vie. If it's an on-purpose thing, then less shadow would be less distracting.

**Comment by ezyang** [comment]
Would me nice if there was some early capsule slide summarizing what interesting new things Pycket does, as opposed to existing JIT systems for other things. Obviously you need to setup some background but I'm still waiting!

**Comment by rrnewton** [comment]
In the optimized dot code do we need to see all the other args to `jump` and `label`?  Did you say what they are?

**Comment by cgswords** [comment]
I'd swap these bleeding reds for another color; at least something a little more subdued? Might have a bad projector at ICFP :/

**Comment by rrnewton** [comment]
@jasonhemann  -- yeah that's a projector thing -- look on the TV screen instead.

**Comment by ezyang** [comment]
(Interpreter-JIT Code-C Code) So, the diagram is confusing: the down arrows seem to be compile, but the leftright are "mode switching"?

**Comment by cfbolz** [comment]
maybe call `f` and `g` something sensible, like `loop` and `myadd`

**Comment by cgswords** [comment]
Could you change the `f`/`g` to, say, the vector example from earlier? Something more concrete?

**Comment by jasonhemann** [comment]
It's a shame there's not a meaningful example that's no more complicated.

**Comment by ezyang** [comment]
It's hard to say what the "essential difficulty" of the example is, compared to Python

**Comment by cgswords** [comment]
"This method is easy to defeat" sounds like you're trying to prevent malicious behavior, not optimize normal code. Which you go back on with the contract thing, so maybe a better way to say this?

**Comment by ezyang** [comment]
I guess the essential difficulty is that we are not writing loops of imperative code, our loops are recursive calls of functions. (OK, finally got to this in "The Callgraph" slide)

**Comment by rrnewton** [comment]
I thought there was supposed to be something about supporting more features than python -- call/cc and contracts.  But I don't think that was emphasized yet.

**Comment by cfbolz** [comment]
@ezyang yes

**Comment by ezyang** [comment]
Would be good if this was made clear earlier :)

**Comment by cfbolz** [comment]
+1

**Comment by cfbolz** [comment]
I wonder whether the whole two-state part is necessary in the talk. I think just describing the callgraph approach is enough, because it's intuitive and satisfying.

**Comment by cgswords** [comment]
Mentioning it in passing might be worthwhile, but yes, it seems the talk would stand fine without it (especially if it's a strategy typical to other rpython programs)

**Comment by cfbolz** [comment]
no, it's not a common technique in rpython (most languages there are imperative and it's completely obvious where the loops are), but it's not a new technique in the tracing space.

**Comment by rrnewton** [comment]
"Data structure Specialization" slide - at 16:30 minutes atm

**Comment by ezyang** [comment]
[Bolz 2013] want to list all authors right?

**Comment by cfbolz** [comment]
In the FloatVectorStrategy picture I wonder whether the 2 should be 2.0, to make it clear that you can't really have ints in there. The paper may have the same problem, but too late for that.

**Comment by rrnewton** [comment]
For all the static-language fans that will be in the audience -- which of these goodies do they not achieve for free in a static language?  Some of the contract checking?  Bounds checking?   A concrete example of this would be nice...

**Comment by cgswords** [comment]
More numbers or something on the axis of the benchmark might be nice? A faint grid maybe?

**Comment by cfbolz** [comment]
@rrnewton actually your talk reminded me a lot of the strategies approach: switch between different representations of a data structure depending on how the data structure is used.

**Comment by ezyang** [comment]
Shootout Benchmarks slide: the changing y-axis scale is confusing.

**Comment by rrnewton** [comment]
"Shootout Becnhmarks" -- it doesn't say how many benchmarks this is taking the geomean over.  Is this the same set that was on the previous slide?  I wasn't paying good enough attention.

**Comment by cfbolz** [comment]
@ezyang agreed, very confusing

**Comment by cgswords** [comment]
I think that was the 20-minute mark, we just got to contracts / chaperones

**Comment by cgswords** [comment]
Yeah changing the axes units for time comparison seems odd / confusing.

**Comment by rrnewton** [comment]
I would nix "For reference".. it's great that it's 2X faster for all benchmarks though ;-).

**Comment by ezyang** [comment]
I missed why in "Benchmark: Contracts" there's a gap in the graph

**Comment by jasonhemann** [comment]
What's the gap in the middle about? For a second I thought it was a benchmark they both go really fast at.

**Comment by cgswords** [comment]
Thank you _after_ the takeaways, not before.

**Comment by samth** [comment]
- Take the victory lap about how it actually is fast on those examples.
- Need to clarify what you're saying on the CEK slide. 
- Too much discussion of how meta-tracing works -- we just need to say that it does.
- Figure out how you will indicate things on the screen when you aren't next to it.
- Make sure to have citations w/ real author lists.
- Using the term "recover performance" is not great.
- Unboxing doesn't just help with numeric benchmarks.
- Spend less time on the strategies slide.
- Don't say "thank you" before the conclusions.

**Comment by ezyang** [comment]
"Virtually eliminate": but in the previous graph you did have a slow down when it was unspecialized?

**Comment by samth** [comment]
Total time: 23:10

**Comment by rrnewton** [comment]
Oops, on the Y axis front... on "Benchmarks: Contracts" I lost track of what "slowdown" was -- percent or times?

**Comment by eholk** [comment]
I like how you explained what the two graphs were showing on the shootouts benchmark. I wasn't sure what you meant by "despecialization" at first, but then it was clear.

End with "Thank you," not thank you and then a summary. Thank you serves as a clear signal to the audience to applaud.

**Comment by rrnewton** [comment]
From Andre: how much of racket is implemented?

Spenser: covering FFI would be hard.  Covered X/Y functions.

**Comment by cfbolz** [comment]
from @rrnewton: why is this harder than PyPy? Python has lambda etc.

**Comment by ezyang** [comment]
samth's restatement of the Q (what does Pycket have which is novel); I would also ask, if you use a lot of lambdas in Python, would it slow down?

**Comment by samth** [comment]
Ryan: what's here that's not in python.

Need to sell the work much harder in response to this question.

**Comment by rrnewton** [comment]
me: what do you give me over python?

spenser: get typed racket, macro system, ... racket may be easier to optimize...

Sam: rephrasing, what technology is here that's not in JITs for other languages (python, lua, Java, etc).

Spenser: here we apply this to a language without explicit loops.  

LuaJit will detect any back-edge in a branch as a loop.  (Does that then cover function calls?   I guess so.)

**Comment by cfbolz** [comment]
the part of the answer that's missing: Python does not do tail call optimization, so recursive functions in PyPy will not be compiled to loops.

**Comment by cgswords** [comment]
Ken is proposing
- extra background: explain back edges, etc.
- use _pictures_ for the call graphs to help with the motivation / explanation
- show expressions evaluated in a call graph
- make a call graph example for the continuation discussion
- make these graphs correspond to the code examples: which 'left parens' correspond to which nodes, etc.
- which criteria give rise to which results on the examples

**Comment by cfbolz** [comment]
could somebody write down Ken's question? I can't hear it on the microphone

**Comment by rrnewton** [comment]
He's saying something about visualizing the callgraph... but I missed part of it because I went to get snacks ;-).

**Comment by cgswords** [comment]
I'm editing them into the comment above

**Comment by samth** [comment]
_Feedback_: @ccshan: had trouble understanding notions of loops. Terminology such as "back edge" confusing. Be more pictorial about the callgraph. `append` example has cycle in callgraph in CPS.

**Comment by cgswords** [comment]
@rrnewton wants to know "Which technology is new" -- and what about call/cc and contract optimizations? Hammering home that novelty is good.

**Comment by samth** [comment]
_Feedback_: @rrnewton: Hinted at the novelty of CEK. Also emphasize novelty of `call/cc` and contracts.

**Comment by vollmerm** [comment]
On the comparison with other Scheme compilers: is this an apples to apples comparison? Like, checking for fixnum overflow, checking bounds on vector references, etc? This is a bit tricky with Scheme compilers, since they don't agree on much.

**Comment by samth** [comment]
@vollmerm Yes, it basically is. Pycket doesn't handle Ctrl-C quite as well as the others, which is a little unfair, but that's it. Fixnums/vectors/etc are all the same.
